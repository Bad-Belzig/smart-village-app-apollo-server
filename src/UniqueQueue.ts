const basicComparison = <T>(a: T, b: T) => a === b;

const isFirstOccurence = <T>(value: T, index: number, self: T[]) =>
  self.indexOf(value) === index;

export class UniqueQueue<T> {
  private queued: T[];
  private completed: T[];
  private duplicateFilterFunc: (value: T, index: number, self: T[]) => boolean;
  private arrayIncludesEntry: (arr: T[], entry: T) => boolean;

  constructor(
    initialValues?: T[],
    duplicateFilterFunc?: (value: T, index: number, self: T[]) => boolean,
    comparisonFunc?: (a: T, b: T) => boolean,
  ) {
    // initialize with empty array or input array filtered for uniqueness
    this.duplicateFilterFunc = duplicateFilterFunc ?? isFirstOccurence;

    const compFunc = comparisonFunc ?? basicComparison;

    this.arrayIncludesEntry = (arr: T[], entry: T) =>
      !(arr.findIndex((arrEntry) => compFunc(entry, arrEntry)) === -1);

    this.queued = initialValues
      ? initialValues.filter(this.duplicateFilterFunc)
      : [];
    this.completed = [];
  }

  // add new entries uniquely to the queued list, of they are not completed already
  add = (...newEntries: T[]) =>
    newEntries.filter(this.duplicateFilterFunc).forEach((entry) => {
      // if the entry is not included in the list of completed entries
      // add it to the queue, and remove duplicates afterwards
      // --> this way the filter function can influence which of the duplicates is removed
      //     (entries can differ but correspond to the same external data point
      //      i.e. for cats we might have one entry { id: ..., weight: ... } and one entry { id: ..., name: ... }
      //      and we can choose to keep the one that has the more relevant data)
      // --> this could be improved in the future to also allow for merging of entries
      if (!this.arrayIncludesEntry(this.completed, entry)) {
        this.queued.unshift(entry);
        this.queued = this.queued.filter(this.duplicateFilterFunc);
      }
    });

  addCompleted = (...newEntries: T[]) => {
    // remove them from being queued if they are flagged as completed
    this.queued = this.queued.filter(
      (entry) => !this.arrayIncludesEntry(newEntries, entry),
    );

    // add them uniquely to the completed entries
    newEntries.filter(this.duplicateFilterFunc).forEach((entry) => {
      if (!this.arrayIncludesEntry(this.completed, entry)) {
        this.completed.push(entry);
      }
    });
  };

  getLength = () => ({
    queued: this.queued.length,
    completed: this.completed.length,
  });

  next = () => {
    const entry = this.queued.pop();
    if (entry) this.completed.push(entry);
    return entry;
  };
}
