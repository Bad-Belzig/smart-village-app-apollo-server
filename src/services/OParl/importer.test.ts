// import { setupTestDataBase } from '../../helpers';
// import { Body, Meeting, System } from '../../models';
import { UniqueQueue } from '../../UniqueQueue';
// import { importOParl } from './importer';

// const helperFuncs1 = setupTestDataBase('test_importer__1_0');
// const helperFuncs2 = setupTestDataBase('test_importer__1_1');

// const testUrl = 'https://www.lwl-pch.sitzung-online.de/oi/oparl/1.0/system.asp';
// const testUrl2 =
//   'https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/system';

describe('the unique queue', () => {
  it('works as expected', () => {
    const queue = new UniqueQueue<string>(
      (value) => value.slice(0, 3),
      () => false,
      ['asds', 'asda'],
    );

    expect(JSON.stringify(queue.getLength())).toEqual(
      JSON.stringify({ queued: 1, completed: 0 }),
    );

    queue.add('asd');
    expect(JSON.stringify(queue.getLength())).toEqual(
      JSON.stringify({ queued: 1, completed: 0 }),
    );

    queue.add('dsa');
    expect(JSON.stringify(queue.getLength())).toEqual(
      JSON.stringify({ queued: 2, completed: 0 }),
    );

    expect(queue.next()).toEqual('dsa');
    expect(JSON.stringify(queue.getLength())).toEqual(
      JSON.stringify({ queued: 1, completed: 1 }),
    );

    expect(queue.next()).toEqual('asds');
    expect(JSON.stringify(queue.getLength())).toEqual(
      JSON.stringify({ queued: 0, completed: 2 }),
    );

    queue.add('asdasdasd');
    expect(JSON.stringify(queue.getLength())).toEqual(
      JSON.stringify({ queued: 0, completed: 2 }),
    );

    queue.add('wasdasdasd');
    expect(JSON.stringify(queue.getLength())).toEqual(
      JSON.stringify({ queued: 1, completed: 2 }),
    );
  });
});

// these tests take too long to be in the CI
// describe('the importer for 1.0', () => {
//   beforeAll(async () => {
//     jest.setTimeout(18000000);
//     helperFuncs1.beforeAll();
//   });

//   it('resolves', async () => {
//     await expect(importOParl(testUrl)).resolves.toBeTruthy();
//   });

//   it('imports a system', async () => {
//     const systemJson = (await System.findOne())?.toJSON();

//     delete systemJson?._id;

//     expect(systemJson).toMatchSnapshot();
//   });

//   it('imports a body', async () => {
//     const bodyJson = (await Body.findOne())?.toJSON();

//     delete bodyJson?._id;

//     expect(bodyJson).toMatchSnapshot();
//   });

//   afterAll(async () => {
//     jest.setTimeout(5000);
//     helperFuncs1.afterAll();
//   });
// });

// describe('the importer for 1.1', () => {
//   beforeAll(async () => {
//     jest.setTimeout(18000000);
//     helperFuncs2.beforeAll();
//   });

//   it('resolves', async () => {
//     await expect(importOParl(testUrl2)).resolves.toBeTruthy();
//   });

//   it('imports a system', async () => {
//     const systemJson = (await System.findOne())?.toJSON();

//     delete systemJson?._id;

//     expect(systemJson).toMatchSnapshot();
//   });

//   it('imports a body', async () => {
//     const bodyJson = (await Body.findOne())?.toJSON();

//     delete bodyJson?._id;

//     expect(bodyJson).toMatchSnapshot();
//   });

//   it('imports a meeting', async () => {
//     const meetingJson = (
//       await Meeting.findOne({
//         externalId:
//           'https://sdnetrim.kdvz-frechen.de/rim4883/webservice/oparl/v1.1/body/1/meeting/486',
//       })
//     )?.toJSON();

//     delete meetingJson?._id;

//     expect(meetingJson).toMatchSnapshot();
//   });

//   afterAll(async () => {
//     jest.setTimeout(5000);
//     helperFuncs2.afterAll();
//   });
// });
