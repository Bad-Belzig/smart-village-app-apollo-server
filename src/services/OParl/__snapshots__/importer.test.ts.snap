// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the filterDuplicateFunction filters as expected 1`] = `
Array [
  Array [
    "asd",
    "Body",
  ],
  Array [
    Object {
      "id": "wasd",
      "other": "asd",
    },
    "Body",
  ],
]
`;
